Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/MixedRealityKeyboardBase.cs

```diff
@@ -0,0 +1,324 @@
+//#define WINDOWS_UWP
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.Events;
+#if WINDOWS_UWP
+using Windows.UI.ViewManagement;
+using Microsoft.MixedReality.Toolkit.Input;
+using System.Collections;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.UI
+{
+    public abstract class MixedRealityKeyboardBase : MonoBehaviour
+    {
+        #region Properties
+
+        /// <summary>
+        /// Returns true if the keyboard is currently open.
+        /// </summary>
+        public bool Visible => state == KeyboardState.Showing;
+
+        /// <summary>
+        /// Returns the index of the caret within the text.
+        /// </summary>
+        public int CaretIndex
+        {
+            get;
+            private set;
+        } = 0;
+
+        [Experimental, SerializeField, Tooltip("Event which triggers when the keyboard is shown.")]
+        private UnityEvent onShowKeyboard = new UnityEvent();
+
+        /// <summary>
+        /// Event which triggers when the keyboard is shown.
+        /// </summary>
+        public UnityEvent OnShowKeyboard
+        {
+            get { return onShowKeyboard; }
```

Comment:
don't block because of this for getting this in. But FYI for future PRs, use the simplified property syntax

```
        public UnityEvent OnCommitText
        {
            get => onCommitText;
            set => onCommitText = value;
        }
```